[ADD]  회원가입 실시간 이메일 중복 체크 기능 구현 (ADD view to check duplicated email)

jhwang0801 opened the pull request:
## :: 최근 작업 주제
- [x] 기능 추가
- [x] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [x] 컨벤션 수정

<br />

## :: 구현 목표 
### - 이메일 중복체크 기능 구현 
    - 회원가입 단계에서 이메일 내용을 작성 시 바로 중복체크를 할 수 있게 하기 위함

<br />

## :: 테스트 코드 결과
![image](https://user-images.githubusercontent.com/75832544/176594293-4459b972-7b5f-482d-9a73-8ae1e037647e.png)

jhwang0801 committed:
ADD view to check duplicated email

- add checking duplicated email function on request
- fix error in WorkingDayView testcode
- fix all conflicts resulting from merged main

jasonkang14 commented on users/views.py:
```diff
@@ -77,4 +77,17 @@ def post(self, request):
+        try:
+            data  = json.loads(request.body)
+            email = data['email']
+            check_duplicate_email(email)
```
음 이거는 저도 자주 실수하는 부분이긴한데(실제로 저희 코드도 이런식으로 되어있구요)
파이썬의 경우 객체지향 언어이기 때문에, 
작성하신 것 처럼 이 view안에서 필요한 유틸리티 함수의 경우 `user.utils`에서 따로 함수를 선언하시는 것 보다는
해당 클래스의 method로 쓰시는 편이 더 좋습니다. 
```python
class CheckDuplicateEmailView(View):
  def check_duplicate_email(self, email):
    ...

  def post(self, request):
    try:
      data = json.loads(request.body)
      email = data['email']
      self.check_duplicate_email(email)
```

그리고 지금 구조는 `check_duplicate_email()`이라는 함수가 리턴하는 값이 없는 것 같은데 에러를 해당 함수에서 발생 시키나요? 

객체지향은 어떤건지 제가 한 번 설명을 드려볼게요!

jasonkang14 replied:
제가 놓쳤던 부분이네요, 
`user.utils`에 있는 `validate_password`나 `generate_jwt`같은 경우도, 해당 함수가 사용되는 view가 하나일 것으로 판단되니
(해당 함수가 사용되는 곳이 여러곳이라면  클래스를 선언하고 상속받아도 되구요)
각각의 view로 옮기는 편이 좋을 것 같습니다. 
여기저기서 사용되는 `login_decorator`만 남겨두는 편이 좋을 것 같습니다

jhwang0801 replied:
말씀해주신 내용 수정하여 반영했습니다!
강의 감사합니다~

jhwang0801 committed:
Improve view and test to be Object-Oriented

- Move validate_password into SignUpView Class
- Add Validation Class
- Move check_duplicates and generate_jwt into Validation Class

jasonkang14 commented on users/tests.py:
```diff
@@ -151,7 +151,7 @@ def test_success_app_patient_login(self):
         self.assertEqual(response.json(), {
             'message'     : 'SUCCESS_PATIENT_LOGIN',
-            'access_token': generate_jwt(user),
+            'access_token': Validation.generate_jwt(user),
```
👍

jasonkang14 replied:
`client = Client()` 와 유사하게
```python
validation = Validation()
validation.generate_jwt(user)
```
가 더 정확한 표현입니다.

하지만 이 테스트 케이스도 클래스니 
```python
class LoginTest(TestCase, Validation):
  self.generate_jwt(user)
```
가 더 좋을 것 같습니다. 
하지만 `generate_jwt`가 `Validation`이 아니라는 생각이 드는데요. 
1. `LoginTest`라는 class의 method로 `generate_jwt`를 선언하거나
2. `LoginView`에서도 사용되니 `Validation`이라는 이름을 바꿔서 `LoginTest`와 `LoginView`가 상속을 받는 것

중에 어떤게 좋아보이는지 재승님의 의견이 궁금합니다!

jhwang0801 replied:
제가 "객체지향"에 대해 이해한 부분이 맞다면 말씀해주신 케이스 중 2번이 더 좋다고 생각합니다.
이유는 `LoginTest`와 `LoginView` 가 하나의 클래스를 상속받지 않게된다면 해당 부분에 수정사항이 생기게 될 때, `LoginTest`와 `LoginView` 모두를 수정해야 올바른 코드(올바르게 작동)가 됩니다. 다르게 얘기하면, 실수 혹은 어떤 이유로 인해 둘 중 하나의 코드만 수정하게 되는 경우는 잘못된 코드가 되버립니다.

결론적으로 하나의 클래스에서 상속받으면(2번 케이스에 해당) 그 클래스만 수정하면 둘 모두에게 반영되어 코드에 오류가 생길 확률이 적어진다 생각합니다.(스파게티 코드를 방지할 수 있게 된다 생각합니다.)

애초에 제가 설정한 컨셉도 2번에 해당합니다.
`LoginView`와 `LoginTest`에서 모두 `generate_jwt`를 사용하기때문에 `Validation`이라는 class안에 method로써 `gnereate_jwt`를 선언해주었습니다. 
(다만 상속에 대해서는 이해가 부족해서 적용하지 못했었습니다. 아래에 리뷰해주신 부분을 보고 명확히 이해가 되었습니다.  :D)

말씀하신대로 `Validation`이라는 단어 자체가 완벽히 적절한 선택은 아니라는 느낌은 받았으나,
한 가지의 method를 선언하기 위해서 하나의 Class를 만드는 것보다는
확인 및 검증의 비슷한 계열 느낌인 `check_duplicate_email` 과 `generate_jwt` method들을 하나의 Class `Validation` 안에 선언해주는 것이 효율적이라고 판단했습니다.

혹은 이런 경우에는 조금 더 명확히 하기위해서 하나의 Method를 위해서 각각 Class를 만들어주는 것이 좋을까요??

jhwang0801 replied:
아!!!! 제가 실수로 Validation 안에도 generate_jwt를 선언하고,
LoginView 안에도 generate_jwt를 선언했네요.....

코드 수정하다 이제서야 발견했습니다. 수정하겠습니다..

jasonkang14 replied:
넵넵 저도 2번이 좋은 것 같습니다! method마다 각각 class를 만드는건 생산성 측면에서 안좋은 것 같아요.

jhwang0801 replied:
넵!! 리뷰 감사합니다!

jasonkang14 commented on users/views.py:
```diff
@@ -20,8 +40,8 @@ def post(self, request):
-            validate_password(password)
-            check_duplicate_email(email)
+            self.validate_password(password)
+            Validation.check_duplicate_email(email)
```
이런 경우에는 아까 말씀드린 상속을 해도됩니다
```python
class SignUpView(View, ValidationView):
  self.check_duplicate_email(email)
```

jhwang0801 replied:
아아..!!! 감사합니다 이해했습니다.

수정해서 반영하겠습니다

jasonkang14 commented on users/views.py:
```diff
@@ -1,16 +1,36 @@
 from users.models import CustomUser
-from users.utils  import check_duplicate_email, validate_password, generate_jwt
+
+class Validation:
```
`views.py`안에는 장고  view만 배치하고, 
기존에 있던 방식대로 `utils.py`에 class를 새로 선언하고, 해당 클래스를 상속받는 편이 더 구조상 좋을 것 같습니다.

jhwang0801 replied:
네!! 수정하도록 하겠습니다!!

jhwang0801 committed:
Improve view and test(2)

jhwang0801 committed:
Improve view and test to be Object-Oriented(3)

jasonkang14 commented on users/tests.py:
```diff
@@ -3,7 +3,7 @@
 
 from users.models import CustomUser
-from users.utils  import generate_jwt
+from users.views  import Validation
```
경로가 잘못되지 않았나요?

jhwang0801 replied:
제대로 확인하지 못하고, 수정하면서 경로 설정을 안바꿔주었습니다.

그런데 정말 신기하게도 테스트코드에서 문제없이 다 돌아가네요...
PUSH하기 전에 몇번씩이나 테스트코드 진행해봤을 때, 전혀 문제가없었고,
지금도 신기해서 테스트코드를 돌려보고있는데 문제없이 돌아가네요....

아마도 제 생각에는 views.py에 `from users.utils import Validation`이 설정되어 있기 때문에  tests.py에서는 `users.views`의 `Validation`을 가져와서 문제없이 테스트코드가 실행되었던 것 같습니다. (`utils.py` ▶︎ `views.py` ▶︎ `tests.py`)

해당 부분은 수정해서 다시 올리도록 하겠습니다!

jasonkang14 replied:
굳굳!

jhwang0801 committed:
Fix the path of Validation Class
